scalanet/ut/src/io/iohk/scalanet/peergroup/DynamicTLSPeerGroupSpec.scala

```diff
@@ -349,6 +375,142 @@ class DynamicTLSPeerGroupSpec extends AsyncFlatSpec with BeforeAndAfterAll {
+      // we are using pretty large messages to be sure incoming and sending buffers will be full
+      val m1 = SizeAbleMessage.genRandomOfsizeN(16 * 1024 * 1024)
+      val m2 = SizeAbleMessage.genRandomOfsizeN(16 * 1024 * 1024)
+      for {
+        finished <- Deferred.tryable[Task, Unit]
+        _ <- clientChannel.sendMessage(m1)
+        _ <- Task.sleep(1.second)
+        // snd buffer and rcv buffer all full, flushing message will be impossible
+        _ <- clientChannel.sendMessage(m2).doOnFinish(_ => finished.complete(())).startAndForget
+        _ <- Task.sleep(2.second)
+        check <- finished.tryGet
+        // even after some time message is not sent as buffers are still full
+        _ <- Task(assert(check.isEmpty))
+        // sending message should finish once receiver will consume event on its side and make space and rcv buffer
+        serverEvent1 <- Task.parMap2(finished.get, serverChannel.nextChannelEvent)((_, sEv) => sEv)
+        serverEvent2 <- serverChannel.nextChannelEvent
+      } yield {
+        assert(serverEvent1.get == MessageReceived(m1))
+        assert(serverEvent2.get == MessageReceived(m2))
+      }
+  }
+
+  it should "get more bytes form rcv buffer as soon as incoming queue will be at least half empty" in backPressureTestCase(
+    clientConfig = getCorrectConfig(maxFrameLength = 64 * 1024 * 1024),
+    serverConfig = getCorrectConfig(maxQueueSize = 4, maxFrameLength = 64 * 1024 * 1024)
+  ) {
+    case (clientChannel, serverChannel) =>
+      val srvChannelImpl =
+        serverChannel.asInstanceOf[DynamicTLSPeerGroupInternals.DynamicTlsChannel[ChannelEvent[SizeAbleMessage]]]
+      // we are using small messages to not fill up snd and rcv buffers
+      val messagesToSend = (0 to 5).map(_ => SizeAbleMessage.genRandomOfsizeN(4 * 1024))
+      val firstBatch = messagesToSend.take(4)
+      val secondBatch = messagesToSend.takeRight(2)
+      for {
+        _ <- Task.traverse(firstBatch)(m => clientChannel.sendMessage(m))
+        _ <- Task(srvChannelImpl.incomingQueueSize).restartUntil(qSize => qSize == firstBatch.size)
+        _ <- Task.sleep(1.seconds)
+        // last two messages will be stored in rcv buffer, becouse incoming queue is full
+        _ <- Task.traverse(secondBatch)(m => clientChannel.sendMessage(m))
+        _ <- Task.sleep(1.seconds)
+        _ <- Task(assert(srvChannelImpl.incomingQueueSize == firstBatch.size))
+        _ <- srvChannelImpl.nextChannelEvent.map(ev => assert(ev.get == MessageReceived(firstBatch(0))))
+        _ <- Task(assert(srvChannelImpl.incomingQueueSize == firstBatch.size - 1))
+        _ <- srvChannelImpl.nextChannelEvent.map(ev => assert(ev.get == MessageReceived(firstBatch(1))))
+        // after reading first 2 messages, there place in queue for two new messages, so they are read from rcv buffer
+        _ <- Task(srvChannelImpl.incomingQueueSize)
+          .restartUntil(qSize => qSize == ((firstBatch.size - 2) + secondBatch.size))
```

Comment:
What does this achieve?

Reply:
Ah, it's like putting `eventually` around the next assertion.

Reply:
If this condition never happened, would the whole test just time out? Maybe it would be worth putting some timeout on this explicitly, so we get a bit more clue about where the test stopped.

Reply:
good idea ! :+1: